stlab/concurrency/default_executor.hpp

```diff
@@ -178,135 +221,188 @@ class task_system {
+
     const unsigned _count{std::thread::hardware_concurrency()};
-    std::vector<std::thread> _threads;
-    std::vector<notification_queue> _q{_count};
+
+    struct thread_context
+    {
+        std::mutex mutex;
+        std::condition_variable ready;
+        std::thread thread;
+    };
+    std::vector<thread_context> _thread_contexts{_count};
+    std::array<std::vector<notification_queue>, 3> _q;
     std::atomic<unsigned> _index{0};
+    std::atomic_bool _done{false};
 
     void run(unsigned i) {
         while (true) {
+            begin:
             task<void()> f;
 
-            for (unsigned n = 0; n != _count * 32; ++n) {
-                if (_q[(i + n) % _count].try_pop(f)) break;
+            for (auto& q : _q) {
+                for (unsigned n = 0; n != 64 / _count; ++n) {
+                    if (q[(i + n) % _count].try_pop(f)) {
+                        f();
+                        goto begin;
+                    };
+                }
+            }
+
+            for (auto& q : _q) {
+                if (q[i].pop(f)) {
+                    f();
+                    goto begin;
+                }
             }
-            if (!f && !_q[i].pop(f)) break;
 
-            f();
+            {
+                lock_t lock{_thread_contexts[i].mutex};
+                while(!_done) {
+                    _thread_contexts[i].ready.wait(lock);
+                    goto begin;
+                }
+            }
+            if (_done) return;
         }
     }
 
 public:
-    task_system() {
+    priority_task_system() {
+        for (auto& q : _q) {
+            std::vector<notification_queue> queues{_count};
+            std::swap(q, queues);
+            for (unsigned n = 0; n != _count; ++n) {
+                q[n].set_context(_thread_contexts[n].ready);
+            }
+        }
         for (unsigned n = 0; n != _count; ++n) {
-            _threads.emplace_back([&, n] { run(n); });
+            _thread_contexts[n].thread = std::thread([&, n] { run(n); });
         }
     }
 
-    ~task_system() {
-        for (auto& e : _q)
-            e.done();
+    ~priority_task_system() {
+        _done = true;
+        for (auto& context : _thread_contexts)
+            context.ready.notify_all();
 
-        for (auto& e : _threads)
-            e.join();
+        for (auto& context : _thread_contexts)
+            context.thread.join();
     }
 
-    template <typename F>
-    void operator()(F&& f) {
+    template <std::size_t P, typename F>
+    void execute(F&& f) {
         auto i = _index++;
 
-        for (unsigned n = 0; n != _count; ++n) {
-            if (_q[(i + n) % _count].try_push(std::forward<F>(f))) return;
+        for (unsigned n = 0; n != 64 / _count; ++n) {
+            if (_q[P][(i + n) % _count].try_push(std::forward<F>(f))) return;
         }
 
-        _q[i % _count].push(std::forward<F>(f));
+        _q[P][i % _count].push(std::forward<F>(f));
     }
 };
 
+inline priority_task_system& pts() {
```

Comment:
This implies `priority_task_system` should not be client-usable by any other means. Should the whole type go into some `detail` or `implementation` namespace?

Reply:
Good point. Will fix this.

Reply:
It just checked. It is already in the detail namespace

Reply:
I think some compilers will create multiple task systems for static locals of inline functions (kinda like a static local of a template function). The ANSI standard says it should only make one but not all compilers do. Is this something we need to account for @FelixPetriconi?